### ISSUE TITLE
Access to HTTP Response Headers of a WebView

### ISSUE THREAD

**Issue by wilsonna** [eclipsesource/tabris-js#638]
It goes like this. I open a URL in a WebView which does authentication. It returns a success page with a token in the response header. Is there anyway to access the response header? I'm open to Custom widgets if that can help. Thanks!

**Comment by mpost** [comment]
Currently that is not possible. We would need to add that as separate feature.

**Comment by wilsonna** [comment]
I see, noted.

**Comment by mpost** [comment]
You wouldn't necessarily need to close the issue. The request is valid.

**Comment by wilsonna** [comment]
Roger that. I managed resolved my problem through other means (external to tabris.js), so it's not crucial to me for now. But it would be really useful for more complex use cases in future.

**Comment by tbuschto** [comment]
Please note that there would be no way to support this in winRT.

**Comment by mpost** [comment]
Reading the headers if also not possible on Android. Only if we were to proxy all requests that go to the webview.

**Comment by wilsonna** [comment]
Unfortunately, this has came back to haunt me. The vendor for the authentication service has changed and my previous workaround no longer works, so now I'm back to having to need access to the response header, namely the Set-Cookie header. It basically boils down to the fact that the WebView does not share the same session as the (tabris.js) application, so even after having authenticated through the WebView, requests sent from the application will fail without the cookie.

I understand the proxy solution won't be coming anytime soon, but I was wondering if anyone has any suggestions as to how I could obtain the HttpOnly cookie? I know this sounds like cookie hijacking, but it's unfortunately how the authentication service is implemented.

**Comment by gosia-eclipsesource** [comment]
@wilsonna Seems like we are just working on the very issue :) The proxy solution might not be that far-fetched after all. You can expect some news in the next days.

**Comment by cookieguru** [comment]
@wilsonna Just a stab in the dark, but is it possible for you to interact with your authentication services without a web view?  In other words, would this work?

1. Create a pair of `TextInput` widgets for the username and password
2. Use [request](https://github.com/request/request) to send that information to the server
3. Read the headers of the response

And based on the text in the Set-Cookie header you can store it in the "session", local storage, etc.

**Comment by wilsonna** [comment]
@cookieguru Thanks for your suggestion. Unfortunately, technical feasibility aside, that would mean the login credentials of the user would have to be made known to the app, which I suspect may be a security violation as the authentication service is a nation-wide single sign-on system to access all government eServices. Furthermore, it also involves 2-factor authentication and a series of redirects, which complicates things further.

**Comment by wilsonna** [comment]
@gosia-eclipsesource I'm really glad to hear that, thanks! :+1:

**Comment by wilsonna** [comment]
@gosia-eclipsesource May I know if there are any estimated time when this will be implemented? Or is there an intermediate solution which I can use? Sorry for bugging.

**Comment by ralfstx** [comment]
@wilsonna Access to HTTP headers won't make it into the 1.8 release, it will be published in 1.9 together with some other new WebView features. However, you can use the experimental property `headers` on WebView to set an object of additional HTTP headers for requests and you should be able to access response headers in the `load` event as `event.headers`.

**Comment by wilsonna** [comment]
@ralfstx Thanks. Sorry if this sounds noobish, but how do I go about testing this experimental property?

**Comment by cookieguru** [comment]
@wilsonna I have not tried it, but this should work:
```js
var webview = new tabris.WebView({
  layoutData: {left: 0, top: 0, right: 0, bottom: 0},
  "url": "http://tabrisjs.com/"
}).on("navigate", function(event) {
  console.log(event.headers);
}).appendTo(page);
```

**Comment by cookieguru** [comment]
Actually, on second look,
@ralfstx Are you sure that's on the `load` event?  According to [the docs](https://tabrisjs.com/documentation/latest/api/WebView) there is no `event` parameter that comes with the load event.  But there is one on `navigate`.

**Comment by cookieguru** [comment]
@wilsonna And in addition to selecting to build against nightly make sure you're pointing to nightly in your `package.json`:
```js
"dependencies": {
  "tabris": "https://tabrisjs.com/downloads/nightly/tabris.tgz"
}
```

**Comment by ralfstx** [comment]
@cookieguru Thanks for providing the snippet.
The headers should be available on `load` event, not `navigate`. The documentation will be adjusted once the API is finalized.

**Comment by wilsonna** [comment]
@ralfstx Sorry for the late reply, only just gotten the opportunity to test this. I updated package.json as described by cookieguru, did an "npm install tabris", verified in node_modules that it's indeed the WebView code that contains the "headers" property, then fired up the Developer app to try it out. However, I'm getting "undefined"  for the console log below. I suppose it would minimally be return an empty object if there are no headers? Is there anything else I should be doing? 

	var webView = tabris.create("WebView", {
		layoutData: {top:0, bottom:0, left:0, right:0},
		url: "http://........"
	}).on("load", function(widget, event) {
		console.log(event.headers);
        });

**Comment by cookieguru** [comment]
The Developer app is running 1.7 of the native code, so while your JS code is listening for the load event and expecting two paramaters, the native code is only going to send one.  The only way to test out this feature would be to build the APK/IPA against nightly **and** include the nightly JS.  I'll have to defer to someone else as to why that's not working.

**Comment by wilsonna** [comment]
@cookieguru Thanks for the explanation. I've done both, so hopefully there's an explanation for this. I'll remain hopeful :pray:

**Comment by ralfstx** [comment]
@cookieguru Sorry for the confusion. Access to headers is only implemented on iOS yet. It turned out that it cannot be implemented in this manner on Android, so the API might change, or, in the worst case, we may even have to drop this feature. /cc @mpost

**Comment by wilsonna** [comment]
That's really bad news for me :disappointed:

**Comment by mpost** [comment]
There is no way to get the response headers from a WebView on Android. There however a few workarounds that can be implemented in Tabris.js. 

1. Use fetch() to get the page and than but the html in the webview.
2. use embedded javascript inside the webview to make the request to the target page and intercept the headers that way.

See http://stackoverflow.com/questions/3134389/access-the-http-response-headers-in-a-webview and

**Comment by wilsonna** [comment]
Unfortunately, the tricky thing about this is that the Response Header can only be read after I've authenticated with the page within the WebView (I'm actually trying to catch the Set-Cookie header). So (1) would not work as there're alot of redirections happening during the authentication process. (2) would also not work as Javascript is not allowed to access the Set-Cookie header. And because it's a httpOnly cookie, I'm also not able to retrieve the cookie with Javascript. I understand that all these limitations on Javascript were enforced for security reasons, to prevent cookie hijacking. I know of other apps who have circumvented this by implementing native code to sniff the cookie at the WebView. But there doesn't seem to be any way for me to work around this issue.

**Comment by cookieguru** [comment]
@wilsonna `fetch` can follow redirects; there's even a module that makes it easy: [go-fetch-js/follow-redirects](https://github.com/go-fetch-js/follow-redirects)

**Comment by wilsonna** [comment]
@cookieguru I'm not sure if follow-redirects works with tabris.js as it has a number of dependencies on node's libraries. Furthermore, aside from redirections, the login page has client side encryption as well as 2FA, so it's not really feasible to "spoof" the login page. As mentioned earlier, this login page is actually a nation-wide single sign-on system which grants access to many government services, as such, the credentials are highly sensitive and I am quite uncomfortable capturing the credentials with a spoofed page and submitting it on-behalf of the original login page.

Thanks for all you patience. If anyone else has any other ideas to share the session (Secure HttpOnly session cookie) between the WebView and the rest of the app, I would really be happy to hear about it. Meanwhile, I'll have to continue cracking my brain over this....

**Comment by wilsonna** [comment]
I'm now considering proxying all my requests through the WebView by appending my JSON into the URL query string. I will then create a servlet to take in these requests, convert them into normal REST requests, listen for the response, then finally redirecting it to a result page with the response JSON in the URL for my app to pickup. I'm guessing this should work.

**Comment by wilsonna** [comment]
I did a POC and in theory, it works. The downside is that I could only execute one query at a time, but that's not such a big problem. The problem is there being a limit to how long the URL can be (in the region of 2000-ish I believe). Even after compressing the JSON, I have a string that's 90,000 characters long, which can't be contained in the URL. Any plans to allow the app to read the body of the WebView? Right now, webview.get("html") only returns the HTML that was set programmatically, and not the contents loaded by the "url" property.

**Comment by wilsonna** [comment]
@mpost Thanks for pointing me in that direction. I suppose I would need to create a custom `WebView` widget? Would appreciate if you could help with the below before I begin:

1. Do I have to implement my own `WebViewOperator` from scratch (like the sample `CalendarOperator`), or is there one which I can extend from?
2. Do I have to upgrade to Pro plan (I'm currently on Developer Legacy) to create a custom widget as stated here https://tabrisjs.com/documentation/latest/custom-widgets-android?

**Comment by gosia-eclipsesource** [comment]
@wilsonna: Regarding 2., you don't need to upgrade. All you need is local build, which is now available for all accounts. Thanks, we will fix that in the docs.

**Comment by wilsonna** [comment]
@gosia-eclipsesource Thanks for the clarification :smile:

**Comment by mpost** [comment]
@wilsonna We have discussed to add api to access the cookies but can't commit to a timeline yet.

Regarding creating your own widget... Yes you could start with a barebone operator like in the calendar example but (if you are adventures) you could also extend the `BrowserOperator` which is internal API. In there you can overwrite the `call` method and listen to your custom calls (like `getCookie`).

From the WebView in js you can write `webView._nativeCall("getCookie", {})`

**Comment by wilsonna** [comment]
@mpost Thanks for the really useful info. I'll explore the BrowserOperator when I have the opportunity. Meanwhile, I'll work on the barebone one first.

**Comment by wilsonna** [comment]
I've created a custom Browser widget (also called BrowserOperator but not extending from the internal API) at https://github.com/wilsonna/tabris-browser2. It's largely copied from the tabris-calendar widget. I'm able to get it to build with my app, but when it tries to create the widget, it'll throw the error `Could not create object "com.wna.Browser" (reason: Could not find IOperationDelegate for null)`. After some trial and error, I'm guessing that my type "com.wna.Browser" is somehow not mapped to the BrowserOperator. Could someone be kind enough to take a look at my code and let me know where I've gone wrong?

**Comment by mpost** [comment]
@wilsonna The `getType()` method of your operator has to return the internal type used in JS: `com.wna.Browser`.

**Comment by mpost** [comment]
Btw... One thing that came to my mind... Since the `CookieManager` is not directly tied to a WebView instance you could actually create a Tabris.js "widget" called `CookieManager`. That will only deal with the cookies instead of the `WebView`.

**Comment by wilsonna** [comment]
@mpost I've actually made some changes since my last post and reverted everything to tabris-calendar, followed by incremental changes to try to identify the problem. The type is now called "ESBrowser" (both in `getType()` and `_type`). Pardon for using the com.eclipsesource package name for now. I'll rename it once I get this to work.

I'm still getting the same error though. I've narrowed the problem down to either the Javascript (looks unlikely since it's largely copied from tabris.js), or the way I implement the BrowserOperator, because the error only occurred when I replaced both of these. i.e. I think it's safe to rule out any configuration issue.

Regarding the CookieManager (I'm using java.net's CookieManager by the way), does that mean I would be able to get all cookies, no just for the app, but even those within the WebViews?

**Comment by mpost** [comment]
I was thinking about the `android.webkit.CookieManager`. I think it shares the cookies between all instances of `android.webkit.WebView` inside one app.

**Comment by wilsonna** [comment]
I started with `android.webkit.CookieManager` as well, but realised it could only return cookies if I were to pass it a URL. I read that `java.net.CookieManager` works the same, but allows listing of all cookies. Unfortunately, I haven't got to the stage to verify this yet.

**Comment by mpost** [comment]
Okay... that took me a while to spot...  :) The BrowserOperator takes either the constructor with `BrowserOperator(TabrisActivity)` or `BrowserOperator(Activity, TabrisContext)`. You are using a constructor `BrowserOperator(TabrisActivity, TabrisContext)`.

**Comment by wilsonna** [comment]
After getting my custom widget to work (`java.net.CookieManager` doesn't work, I had to revert to `java.net.CookieManager`), I later discovered that the Cordova Plugin CookieMaster (https://github.com/kristianhristov/cordova-cookie-master) actually works equally well, and it supports iOS too! :tired_face: I'm not sure why it didn't work the first time I tried it, but I'm sure glad it works now! I'll just put this here in case anyone needs it in future.

`config.xml`

    <plugin name="com.cordova.plugins.cookiemaster" />

And in your js.

	tabris.create("WebView", {
		layoutData: {top:0, bottom:0, left:0, right:0},
		url: 'http://your.login.url'
	}).on("load", function(widget, event) {
		var url = widget.get("url");
		if (url.indexOf('a-string-in-the-url-that-signals-login-successful') >= 0) {
			cookieMaster.getCookieValue(url, 'MyCookieName', function(data) {
				var cookieValue = data.cookieValue;
				// do whatever you want with the cookie value 
			}, function(error) {
				if (error) {
					console.log('error: ' + error);
				}
			});
		}
	}).appendTo(page);

It wasn't the end of the story for me at this point, because in order to pass this cookie value in my XHR, I had to override the default `XMLHttpRequest` that came with tabris.js, to remove the restriction on setting the `Cookie` request header.

Thanks @mpost for your patience in helping me out. At least now, I know a little about creating custom widgets! :smile:

**Comment by mpost** [comment]
@wilsonna Glad that you managed to bring this functionality to work. We realize that reading the cookie store of an application is very valuable for login scenarios and we are working on a "native" implementation ourselfs. In the meantime it is good to hear that the plugin workes for you. :)

As you have seen we are also working on relaxing the XHR restrictions to be more usable in the Native app scenario. https://github.com/eclipsesource/tabris-js/issues/904

**Comment by wilsonna** [comment]
Just another update for the benefit of anyone who's interested.

While cookie-master works in retrieving the cookie from the WebView for Android, it does not work for iOS because I believe tabris.js uses the WKWebView (introduced from iOS 8 onwards) for its WebView, while cookie-master only works for UIWebView (introduced since iOS 2). It's close to impossible to get the cookies out of WKWebView as there is no API to do so.

Instead, I had to use the Cordova InAppBrowser in place of tabris.js's WebView, since it uses the UIWebView and works nicely with cookie-master.

I'm now happy to report that I was finally able to retrieve the session cookie which I needed for both Android and iOS and my authentication problems are truly and finally over :smile: 

But of course, a native implementation would be very much preferred, but I foresee some difficulties with WKWebView, as I've experienced.
